### PR TITLE
Clear snapshot from prefetch cache after 5s or on load

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Clear snapshot from prefetch cache after 5s or on load. If we clear it immediately then we will make a network call again to fetch this snapshot which is redundant. So don't clear this cache immediately. We also clear it once it is used.

### CODE REVIEW

**Comment by vladsud** [comment]
@jatgarg , could you please add some description on why this change is needed? I.e., what problem it solves? It's so hard to guess. :)

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 1]
```diff
@@ -437,6 +437,8 @@ export class OdspDocumentStorageService extends OdspDocumentStorageServiceBase {
 			const result = await this.cache.snapshotPrefetchResultCache
 				?.get(prefetchCacheKey)
 				?.then(async (response) => {
+					// Remove it from cache once used.
+					this.cache.snapshotPrefetchResultCache.remove(prefetchCacheKey);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:441]: prefetchLatestSnapshot() that puts this cache entry will delete it as well. It feels like this should not be required, no?
I.e. what is the scenario where removing it here is required / changes the flow?

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:441]: @vladsud I described in the PR description and by comment in the prefetchSnapshot file. 
Scenario is that if we clear the cache immediately after caching it in persistent cache, then during load, when we fetch the snapshot parallely, then we won't find the snapshot in non persistant cache and make the network call again. So we will be making an extra Trees latest call. So that is why we are clearing it after 5s or once it used on load.

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts [hunk 2]
```diff
@@ -141,15 +141,19 @@ export async function prefetchLatestSnapshot(
 					});
 					assert(cacheP !== undefined, 0x1e7 /* "caching was not performed!" */);
 					await cacheP;
+					// Schedule it to remove from cache after 5s, so that if this snapshot is going to be used on load,
+					// and we don't find it in this cache, then we will make a network call again to fetch this
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:145]: It might be hard to follow (for someone not knowing details of snapshot fetching flow).
Maybe
```
Schedule it to remove from cache after 5s.
1. While it's in snapshotNonPersistentCache: Load flow will use this value and will not attempt to access persistent cache or fetch snapshot from network. That's the best from perf POV, but cache is not being updated, so it gets stale after a while, thus we want to eventually remove snapshot from this cache.
2. After it's removed from snapshotNonPersistentCache: snapshot is present in persistent cache, so we sill still use it (in accordance with cache policy controlled by host). But load flow will also fetch snapshot (in parallel) from storage and update cache. This is fine long term, but is an extra cost (unneeded network call).
```

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts [hunk 3]
```diff
@@ -141,15 +141,19 @@ export async function prefetchLatestSnapshot(
...
+					// snapshot which is redundant. So don't clear this cache immediately. We also clear this cache,
+					// if this cached snapshot is used on load.
+					setTimeout(() => {
+						snapshotNonPersistentCache?.remove(nonPersistentCacheKey);
+					}, 5000);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:150]: Given that we remove this same cache line in fetchSnapshotCore(), should this number be bigger (like 10 seconds)?
I.e. we expect that in all cases if load follows (that is 99% case), then we use it and delete outside of the timer flow. And Making this longer ensures more sessions would hit it?

On other side, we do not cancel this timer on fetchSnapshotCore() (it would be hard to do), thus it could theoretically run into problems with multiple load sequences, so maybe keeping it low is beneficial? Not sure.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts:150]: I think we can be data driven here in immediate term and see how 5s works out. I don't expect many containers to load from same snapshot and I also don't expect apps to make prefetch call like more than 5s ago before they load the container. So for now 5s seems fine to me.
